projects/baby_driver/test/test_i2c_write.c

```diff
@@ -0,0 +1,247 @@
+  TEST_ASSERT_EQUAL(BABYDRIVER_MESSAGE_STATUS, s_received_data[0]);
+  TEST_ASSERT_EQUAL(s_status_return, s_received_data[1]);
+
+  // Invalid babydriver_id for data messages
+  uint8_t inv_2[8] = { BABYDRIVER_MESSAGE_I2C_WRITE_COMMAND, 0, 0, 0, 0, 0, 0, 0 };
+
+  CAN_TRANSMIT_BABYDRIVER(command[0], command[1], command[2], command[3], command[4], command[5],
+                          command[6], command[7]);
+
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+  MS_TEST_HELPER_ASSERT_NO_EVENT_RAISED();
+
+  CAN_TRANSMIT_BABYDRIVER(inv_2[0], inv_2[1], inv_2[2], inv_2[3], inv_2[4], inv_2[5], inv_2[6],
+                          inv_2[7]);
+
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+
+  TEST_ASSERT_EQUAL(2, s_times_callback_called);
+  TEST_ASSERT_EQUAL(BABYDRIVER_MESSAGE_STATUS, s_received_data[0]);
+  TEST_ASSERT_EQUAL(s_status_return, s_received_data[1]);
+
+  // Invalid port for command message
+  uint8_t inv_3[8] = { BABYDRIVER_MESSAGE_I2C_WRITE_COMMAND, 2, 0, 0, 0, 0, 0, 0 };
+
+  CAN_TRANSMIT_BABYDRIVER(inv_3[0], inv_3[1], inv_3[2], inv_3[3], inv_3[4], inv_3[5], inv_3[6],
+                          inv_3[7]);
+
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+
+  TEST_ASSERT_EQUAL(3, s_times_callback_called);
+  TEST_ASSERT_EQUAL(BABYDRIVER_MESSAGE_STATUS, s_received_data[0]);
+  TEST_ASSERT_EQUAL(s_status_return, s_received_data[1]);
+
+  // Invalid tx_len for command message
+  uint8_t inv_4[8] = { BABYDRIVER_MESSAGE_I2C_WRITE_COMMAND, 0, 0, 6, 0, 0, 0, 0 };
+
+  CAN_TRANSMIT_BABYDRIVER(inv_4[0], inv_4[1], inv_4[2], inv_4[5], inv_4[4], inv_4[5], inv_4[6],
+                          inv_4[7]);
+
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+
+  TEST_ASSERT_EQUAL(4, s_times_callback_called);
+  TEST_ASSERT_EQUAL(BABYDRIVER_MESSAGE_STATUS, s_received_data[0]);
+  TEST_ASSERT_EQUAL(s_status_return, s_received_data[1]);
+}
+
+// Test whether timeout error is called when no message is returned
+void test_timeout_error_i2c(void) {
+  s_status_return = STATUS_CODE_TIMEOUT;
+
+  // Test whether callback will timeout when no data is received after command message
+  uint8_t command[8] = { BABYDRIVER_MESSAGE_I2C_WRITE_COMMAND, I2C_PORT_1, 10, 7, 0, 0, 0, 0 };
+
+  CAN_TRANSMIT_BABYDRIVER(command[0], command[1], command[2], command[3], command[4], command[5],
+                          command[6], command[7]);
+
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+
+  TEST_ASSERT_EQUAL(1, s_times_callback_called);
+  TEST_ASSERT_EQUAL(BABYDRIVER_MESSAGE_STATUS, s_received_data[0]);
+  TEST_ASSERT_EQUAL(s_status_return, s_received_data[1]);
+}
```

Comment:
Nice test for the timeout condition! So that this test doesn't take 750ms to run, consider making the timeout configurable via passing a `uint32_t` into `i2c_write_init` and storing it in a static variable. Then you can pass something small in the tests so it doesn't take a long time to run, and pass the normal timeout in main.c.